### ISSUE TITLE
how to accelerate bert classification model with tensorrt ?

### ISSUE THREAD

**Issue by pogevip** [NVIDIA/TensorRT#931]
how to accelerate bert classification model with tensorrt ?

**Comment by ttyio** [comment]
Hello @pogevip , thanks for question.
We have a BERT SQuAD sample in https://github.com/NVIDIA/TensorRT/tree/release/7.2/demo/BERT
Can you take this as example? Thanks!

**Comment by TLCFYBJJHYYSND** [comment]
the output of TensorRT in SQUAD sample was batch_size*seq_length*2，what should i do if i want get the hidden_state of CLS to do classification task.

**Comment by ttyio** [comment]
Hello @TLCFYBJJHYYSND , I have not try, you can read https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/modeling.py#L126,
replace the last squad layer in `builder.py` with `pooler` layer, then append classifier layer (`matmul` + `softmax`) to change the network to classification.

**Comment by TLCFYBJJHYYSND** [comment]
@ttyio 
Thank you! i will try it soon

**Comment by pogevip** [comment]
@ttyio

**Comment by pogevip** [comment]
@ttyio I've been studying tensorrt recently. 
https://github.com/NVIDIA/TensorRT/blob/master/demo/BERT/builder.py
I use builder.py to convert my tf model to engine.  But when I use the Torch Model or the TFModel whose weight name is different from builder.py, do I need to rewrite builder.py load_tf_weights?

**Comment by pogevip** [comment]
> 你好@TLCFYBJJHYYSND，我还没有尝试，你可以阅读https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/modeling.py#L126，
> 在替换最后一个小队层`builder.py`与`pooler`层，然后附加分类器层（`matmul`+ `softmax`）以将网络更改为分类。
As i know, this script implements multi-GPU multi-precision training model with TensorFlow, but does not mention TensorRT?

**Comment by pogevip** [comment]
> 你好@TLCFYBJJHYYSND，我还没有尝试，你可以阅读https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/modeling.py#L126，
> 在替换最后一个小队层`builder.py`与`pooler`层，然后附加分类器层（`matmul`+ `softmax`）以将网络更改为分类。

据我所知，这个脚本工具多GPU多精度训练TensorFlow模型，但没有提到TensorRT？

**Comment by pogevip** [comment]
> Hello @TLCFYBJJHYYSND , I have not try, you can read https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/modeling.py#L126,
> replace the last squad layer in `builder.py` with `pooler` layer, then append classifier layer (`matmul` + `softmax`) to change the network to classification.

As i know, this script implements multi-GPU multi-precision training model with TensorFlow, but does not mention TensorRT?

**Comment by ttyio** [comment]
> @ttyio I've been studying tensorrt recently.
> https://github.com/NVIDIA/TensorRT/blob/master/demo/BERT/builder.py
> I use builder.py to convert my tf model to engine. But when I use the Torch Model or the TFModel whose weight name is different from builder.py, do I need to rewrite builder.py load_tf_weights?

Yes, currently the sample provide `load_onnx_weights_and_quant` and `load_tf_weights`, and you can override with your own loader.

**Comment by ttyio** [comment]
> > Hello @TLCFYBJJHYYSND , I have not try, you can read https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/modeling.py#L126,
> > replace the last squad layer in `builder.py` with `pooler` layer, then append classifier layer (`matmul` + `softmax`) to change the network to classification.
> 
> As i know, this script implements multi-GPU multi-precision training model with TensorFlow, but does not mention TensorRT?

Yes, the link provide here only as example on how to use BERT in a classification task. It is not TRT based.

**Comment by pogevip** [comment]
> > > 你好@TLCFYBJJHYYSND，我还没有尝试，你可以阅读https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/modeling.py#L126，
> > > 在替换最后一个小队层`builder.py`与`pooler`层，然后附加分类器层（`matmul`+ `softmax`）以将网络更改为分类。
> > 
> > 
> > 据我所知，该脚本使用TensorFlow实现了多GPU多精度训练模型，但没有提到TensorRT吗？
> 
> 是的，此链接仅作为示例提供有关如何在分类任务中使用BERT的示例。它不是基于TRT的。

Thank you!!

**Comment by TLCFYBJJHYYSND** [comment]
@pogevip 
Wonder that whether tf1 or tf2 you are working with?
I find out that the the builder.py cannot wok on tf2

**Comment by ttyio** [comment]
@TLCFYBJJHYYSND try

    import tensorflow.compat.v1 as tf
    tf.disable_v2_behavior()

**Comment by TLCFYBJJHYYSND** [comment]
感谢回答，我遇到的问题是，tf2训练后保存的模型的参数和tf1的有所不同，所以在例如`trt.PluginFieldCollection`这些的时候，无法像bert demo里的builder一样的引用这些参数，导致模型无法转换成engine的形式，请问这有什么解决办法吗

**Comment by ttyio** [comment]
Hello @TLCFYBJJHYYSND , 
Do you mean the key in the checkpoint's weights dict changed?
If you you need add the mapping conversion code in the `load_tf_weights` in builder.py, the checkpoint here is trained from https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/

**Comment by TLCFYBJJHYYSND** [comment]
@ttyio 
Yes ， that's what i'm talking about, the key of dict and the shape of weights corresponding have changed. I will try to add some conversion code. Thanks again.
